core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -665,36 +644,54 @@ public InputStream openExistingFile(String path, OpenOptions options) throws IOE
 
   @Override
   public boolean renameDirectory(String src, String dst) throws IOException {
+    if (exists(dst)) {
+      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
+      return false;
+    }
+    // Use a global delete buffer, in order to merge delete object requests
+    DeleteBuffer deleteBuffer = new DeleteBuffer();
+    boolean result = renameDirectoryInternal(src, dst, deleteBuffer);
+    int fileDeleted = deleteBuffer.getResult().size();
+    if (fileDeleted != deleteBuffer.mEntriesAdded) {
+      LOG.warn("Failed to rename directory, successfully deleted {} files out of {}.",
+          fileDeleted, deleteBuffer.mEntriesAdded);
+      return false;
+    }
+    return result;
+  }
+
+  private boolean renameDirectoryInternal(String src, String dst, DeleteBuffer deleteBuffer)
+      throws IOException {
     UfsStatus[] children = listInternal(src, ListOptions.defaults());
     if (children == null) {
       LOG.error("Failed to list directory {}, aborting rename.", src);
       return false;
     }
-    if (exists(dst)) {
-      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
-      return false;
-    }
     // Source exists and is a directory, and destination does not exist
     // Rename the source folder first
-    if (!copyObject(stripPrefixIfPresent(convertToFolderName(src)),
-        stripPrefixIfPresent(convertToFolderName(dst)))) {
+    String srcKey = stripPrefixIfPresent(convertToFolderName(src));
+    if (!copyObject(srcKey, stripPrefixIfPresent(convertToFolderName(dst)))) {
       return false;
     }
+    deleteBuffer.add(srcKey);
```

Comment:
if renameDirectoryInternal gets called recursively does that mean at certain point the batch threshold is reached and we started a thread to delete before we can do renaming?

Reply:
In the object store, the move operation is simulated by two operations: copying and deleting objects. We will copy and then delete objects in parallel, and when these operations are done, the move operation is completed.

Reply:
so u r saying if srckey is a f1/ with thousands subdir and subfiles inside (e.g. so it has f1/f11/ and f1/obj1, one folder one obj), this copyobj op, say from f1/ to f2/ will only rename the folder itself? and we rely on the next rename section to rename f1/f11/ to f2/f11 and f1/obj1 to f2/obj1?

Reply:
right.